_posts/blog/2016-07-18-subjects.md

```diff
@@ -0,0 +1,142 @@
+
+{% highlight java %}
+PublishSubject<String> subject = PublishSubject.create();
+subject.subscribe(text -> Log.d(“Subjects:”, text));
+subject.onNext(“Easy to use”);
+{% endhighlight %}
+
+This is because ``Subjects`` extend ``Observable`` and implement ``Observer``. 
+
+Thanks to being an ``Observable``, we can use all the typical operators (``map``, ``flatMap``, etc) without any extra hassle:
+
+{% highlight java %}
+subject.filter(str -> !str.isEmpty())
+.map(str -> “This is a not empty string: ” + str);
+{% endhighlight %}
+
+The benefit of being an ``Observer`` is that we can directly put any value to ``onNext``, throw an exception in ``onError`` or just complete. 
+
+{% highlight java %}
+subject.onNext(“Some value”);
+subject.onComplete();
+{% endhighlight %}
+
+## ... comes with a price
+
+Some say that ``Subjects`` are broken, but I would disagree. The creators of ``Subjects`` did a great job, they've focused on creating a user-friendly tool, despite some limitations.
+
+With ``Subjects``, a great power comes with a great burden.
+
+### Backpressure
+
+Basically, backpressure exception occurs when an operator produces values faster than you can handle further downstream. 
+
+I like to compare operators that can handle backpressure to waiters. If clients of a restaurant are full, then they would tell the waiter to stop bringing food and naturally, no more food will be served. On the other hand, operators that cannot handle backpressure are like grandmas. If you tell her that you cannot eat any more, she will just ignore you and keep on feeding you more food until you explode or run away. 
+
+
+The issue with ``Subjects`` and backpressure comes from the fact that ``Subjects`` do not extend ``Subscriber``. As <a href="https://twitter.com/akarnokd">Dávid Karnok</a> mentioned in his blog, ``Subscriber`` is the main way to handle backpressure in RxJava. Unfortunately, Java does not have multiple inheritance. The designers have chosen to make ``Subjects`` simpler to use rather than support backpressure. 
+
+There is really simple test to check if your ``Observable`` supports backpressure. The trick is to request via ``TestSubscriber`` less elements that we put into the ``Subject``:
+
+{% highlight java %}
+@Test
+public void check_backpressure_support() {
+  TestSubscriber<Boolean> ts = new TestSubscriber<>(**1**); // The Subscriber requests only 1 value
+  PublishSubject<Boolean> subject = PublishSubject.create();
+  subject.subscribe(ts);
+
+  subject.onNext(true); // We put 2 values into the stream
+  subject.onNext(true);
+
+  ts.assertValueCount(**1**); // We should receive only 1 value
+}
+{% endhighlight %}
+
+Let’s assume ``ts.assertValueCount(1)`` passes, that would mean that subject supports backpressure. But ``PublishSubject`` does not support backpressure, so in this fails as there will be two values emitted by the ``Subject``.
+
+Solution:
```

Comment:
if you want to have "Solution:" on a separate line you need to add one more new line after it.